### ISSUE TITLE
Storage missing

### ISSUE THREAD

**Issue by MovGP0** [dotnet/TorchSharp#581]
In torch there is the `storage()` method to get a `Storage` object that acts as a proxy to access the underlying data array in a linear fashion. 

**Example**
```csharp
var matrix = tensor(new double[] { 0, 1, 2, 3, 4, 5 }, new long[] {2L, 3L});
// matrix is now [[0, 1], [2, 3], [4, 5]]

var storage = matrix.storage();
storage[3] = 6;
// matrix is now [[0, 1], [2, 6], [4, 5]]
```

**See also**
* [Tensor.storage() method](https://pytorch.org/docs/stable/generated/torch.Tensor.storage.html)
* [Storage class](https://pytorch.org/docs/stable/storage.html)

**Comment by MovGP0** [comment]
**Remark** 
It might be tricky to implement the setters properly, because the getter returns a point object in python: 

```python
matrix = torch.tensor([[4.0, 1.0], [5.0, 3.0], [2.0, 1.0]])
point = matrix[3]
matrix.storage_offset() # returns: 4
```

**Comment by NiklasGustafsson** [comment]
Yes, Storage is one of the low-level features that has taken a backseat to other things. It's on the backlog now.

**Comment by NiklasGustafsson** [comment]
BTW, @MovGP0, I'm curious about the use cases for Storage -- is it just linear indexing? What can you do with Tensor.storage() that you cannot do with 'Tensor.data()'?

Thanks,

Niklas

**Comment by NiklasGustafsson** [comment]
> **Remark** the getter returns a Tensor object in python:

It looks like something else than a Tensor when I try:

```
>>> t.storage()
 1.0
 1.0
 1.0
 1.0
[torch.FloatStorage of size 4]
>>>
```

**Comment by NiklasGustafsson** [comment]
There will be a couple of minor challenges with mapping the storage functionality to the .NET static type system:

1.  'storage()' may return one of several classes, each corresponding to the element type. The way I was planning on mapping this to .NET was to declare 'storage()' to return a non-generic base class, while the dynamic type is Storage<T>, where T is the .NET element type. This means that there will be some type-casting before you can use the Storage instance to get to data. For example:

```csharp
            var st = (Storage<long>)x.storage();
```

2. When passing in a range, the index getter returns another Storage<T> object. However, the setter only accepts a scalar, even for slices. This means that the indexer taking a System.Range will be declared to return/take 'object'. It means that the returned result has to be type-cast, __and__ it means that you have to have the exact right type when setting. For example, an 'int' won't convert to 'long' automatically:

```csharp
var data = new long[] { 1, 1, 2, 2, 3, 1, 1, 2 };
var x = torch.tensor(data);
var st = (Storage<long>)x.storage();
var sl0 = (Storage<long>)st[(3, 5)];
st[(3,5)] = 5L;
```

**Comment by NiklasGustafsson** [comment]
>> Also, why the indexer of Storage<long> doesn't return Storage<long> and need a cast?

Because the setter will take 'long' and the getter with return Storage<long>. 'object' is the only common base type.

**Comment by tarekgh** [comment]
>  the setter will take 'long' and the getter with return Storage. 'object' is the only common base type.

Sorry if I am missing anything here as I do not fully know how things internally work. 
If we'll have `Storage<T>`, why can't we have the indexer setter takes `T` and the getter return `Storage<T>`?

**Comment by NiklasGustafsson** [comment]
Because the way setters and getters are declared for properties (which indexers are), there's only one property type.

**Comment by NiklasGustafsson** [comment]
>> Can't we make the storage() a generic method? 

That's an interesting idea. I wonder what we should do if the "T" type argument doesn't correspond to the element type of the tensor -- raise an exception, or convert (since Storage instances can be converted to other element types).

**Comment by tarekgh** [comment]
> raise an exception, or convert (since Storage instances can be converted to other element types).

Both options are viable. I prefer the conversion if we can implement it correctly. I mean the conversion will not cause data trimming or overflows.

**Comment by NiklasGustafsson** [comment]
>> I mean the conversion will not cause data trimming or overflows.

It would have the same semantics as explicit conversion, which would do whatever the Torch runtime does when converting tensor data.

In other words: 

```csharp
var t = torch.tensor(new long[] { ... });
var st0 = t.storage<byte>();
var st1 = t.storage<long>().byte();
```
would be equivalent.

**Comment by tarekgh** [comment]
Conversion looks like a good idea to me then. 

Just for my education, when doing `t.@byte()` does it trim the tensor values today?

**Comment by NiklasGustafsson** [comment]
It rounds and then strips off the most-significant bits.

**Comment by luisquintanilla** [comment]
I like the look of a generic storage method compared to the casting approach. 

Re: conversion, I assume Torch already does that since there's only a single `storage` method. If it's something users already expect, conversion sounds good.

**Comment by NiklasGustafsson** [comment]
> I like the look of a generic `storage` method compared to the casting approach.
> 
> Re: conversion, I assume Torch already does that since there's only a single `storage` method. If it's something users already expect, conversion sounds good.

No, torch looks at the element type and returns an instance of the corresponding storage type, so there's no conversion unless you then call '.byte/short/int/...()' explicitly. The issue with .NET is that you might specify a 'T' that is not the same as the element type.

**Comment by NiklasGustafsson** [comment]
>> I like the look of a generic storage method compared to the casting approach.

I do, too, by the way.

**Comment by NiklasGustafsson** [comment]
>> When passing in a range, the index getter returns another Storage object. 

We could, perhaps, have an implicit conversion operator that takes a 'T' and converts to Storage<T> containing one element, and sets a flag that marks it as a scalar Storage<T>, which the indexed setter then looks for. It feels a bit hacky though.

**Comment by tarekgh** [comment]
> there's no conversion unless you then call '.byte/short/int/...()' explicitly. The issue with .NET is that you might specify a 'T' that is not the same as the element type.

You can constraint the generic type to the types you want support. Here is an example of how to do that https://github.com/dotnet/runtime/blob/0a2d2548e8a1f8d075529555e6a7e70032c7d9f0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs#L118

**Comment by NiklasGustafsson** [comment]
Sure, it's already constrained to 'unmanaged,' but that's not the problem -- 'T' is a static type, specified in the source code, while the tensor element type is dynamic. So, there's nothing that prevents you from calling 'storage<double>()' on a int tensor. In Pytorch, you don't have that problem, the returned instance will always correspond to the element type.

**Comment by tarekgh** [comment]
What I was trying to say is we support only specific types we know. I mean `T` will be `byte`, `short`, `long`,...etc. Inside the `Storage<T>()` we manually check the generic type and return the matched result. For example, when calling `Storage<byte>()`, the implementation will manually call `torch.@byte()`.

**Comment by NiklasGustafsson** [comment]
Yes,

**Comment by MovGP0** [comment]
> 

I already did. The PR Seems fine to me.

**Comment by NiklasGustafsson** [comment]
Storage was implemented in #587
